### DISCUSSION TITLE
IS anyone else having issues with floats unwantedly turning into complex numbers?  1.0 => 1j ?

### DISCUSSION THREAD

**Discussion by gitcnd** [micropython/micropython#13447] [ESP32]
I'm doing stuff like this:

```
tin = 1.0
print(tin)
```

and the output is this

`1j`

(plus trying to use it in time.sleep(tin) whines about not allowing complex numbers).

I'm on bleeding-edge: MicroPython v1.23.0-preview.43.g215a982c1

**Comment by robert-hh** [comment]
No, not an any ESP32 variant. I assume that you use an ESP32. If it is repeatable, it could be a problem of your terminal software or your board.

**Comment by peterhinch** [comment]
If the firmware is really doing this it would break any application that used floats. Are you sure your firmware build matches your ESP32 hardware?

Please describe the ESP32 hardware you're using if you want us to try to replicate this problem.

**Comment by Josverl** [comment]
FWIW: `float : 1.0` is covered in the test suite in multiple places such as : [tests/float/builtin_float_abs.py](https://github.com/micropython/micropython/blob/16c6bc47cf2b9247418bca8c0d1361167d08594e/tests/float/builtin_float_abs.py)

So if it is actually a bug, the errors should show up in the test results of that build, although I'm not sure in which of the CI steps that could be found

**Comment by sosi-deadeye** [comment]
Maybe you get `complex` by accident.

```python
>>> (-1) ** 0.5
(6.123233995736766e-17+1j)
```

Run your code and if it still produces the imaginary number, then post your code.
I guess something with the code is wrong.

**Comment by gitcnd** [comment]
This is the code:

```
# blink.py

__version__ = '1.0.0' # Major.Minor.Patch

from sh import human
import os
import machine
import time


def __main__(args):
        num = 10
        pin=2
        if len(args) > 2:
                pin = int(args[2])
        if len(args) > 3:
                num = int(args[3])
        ton = 1.0
        if len(args) > 4:
                ton = float(args[4])
        toff = 1.0
        if len(args) > 5:
                toff = float(args[5])

        if not num:
                num = 10

        print("blink pin{} loop{} ondelay{} offdelay{}".format(pin, num, ton, toff))
        led = machine.Pin(pin, machine.Pin.OUT)

        while num>0:
            led.value(1)
            time.sleep(ton)
            led.value(0)
            time.sleep(toff)
            num=num-1
```

The line that fails with complex number error is time.sleep(ton)
The "print" line outputs this:-

blink pin2 loop10 ondelay1j offdelay1.0

globally renaming "ton" to "tion" fixed the problem... which indicates to me that some kind of weird stability issues must be at play with either my build, or my board.

The same 1j problem appeared in a different program of mine as well.

What makes *no* sense, is that there's a LOAD of other python stuff running before mine does, all of which is fine.

It's an ESP32CAM board.

**Reply by robert-hh** [reply]
What is `sh` doing and how did you call blink.py when it fails?

**Comment by stinos** [comment]
Can you print the value of `args` as well, otherwise it's hard to figure out what is wrong.

Note you should be able to `print(f"{args=}")` and `print(f"{pin=} {num=}")` etc

> there's a LOAD of other python stuff running before mine does

But is this sample minimal, i.e. is this the *only* thing needed to reproduce the issue?
